pytoda/datasets/_polymer_dataset.py

```diff
@@ -0,0 +1,360 @@
+        """Set dataset to return the original SMILES strings (wiht the sart-end
+        tokens) intead of tensors
+        """
+        if not self._returning_tensors:
+            return
+        self._bakup_transforms = []
+
+        for ds in self._datasets:
+            self._bakup_transforms.append(deepcopy(ds._dataset.transform))
+            ds._dataset.transform = Compose(
+                [
+                    self.smiles_language.add_start_stop_tokens,
+                    lambda x: ''.join(x)
+                ]
+            )
+        self._returning_tensors = False
+
+    def set_mode_tensor(self):
+        """Set dataset to return tensors of tokens"""
+        if self._returning_tensors:
+            return
+        for ds, transform in zip(self._datasets, self._bakup_transforms):
+            ds._dataset.transform = transform
+        self._returning_tensors = True
+
+    def __len__(self, *args, **kwargs):
+        raise NotImplementedError()
+
+    def __getitem__(self, *args, **kwargs):
+        raise NotImplementedError
+
+
+class _PolymerDatasetAnnotation(_PolymerDataset):
+
+    def __init__(self, *args, **kwargs):
+        # Set `use_anotation_files` to True
+        super().__init__(True, *args, **kwargs)
+
+    def __len__(self) -> Iterable[int]:
+        """Total number of samples."""
+        return self.number_of_samples
+
+    def __getitem__(self, index: int) -> Iterable[torch.tensor]:
+        """
+        Generates one sample of data.
+
+        Args:
+            index (int): index of the sample to fetch.
+
+        Returns:
+            Tuple: a tuple containing self.entities+1 torch.Tensors
+            representing respetively: compound token indexes for each chemical
+            entity and the property labels (annotations)
+        """
+
+        # sample selection
+        selected_sample = self.annotated_data_df.iloc[index]
+
+        # labels (annotations)
+        labels_tensor = torch.tensor(
+            list(selected_sample[self.labels].values),
+            dtype=torch.float,
+            device=self.device
+        )
+        # samples (SMILES)
+        smiles_tensor = tuple(
+            map(
+                lambda x: x[
+                    x.sample_to_index_mapping[selected_sample[x.name]]
+                ],
+                self._datasets
+            )
+        )  # yapf: disable
+        return tuple([*smiles_tensor, labels_tensor])
+
+
+class _PolymerDatasetNoAnnotation(_PolymerDataset):
+
+    def __init__(self, *args, **kwargs):
+        # Set `use_anotation_files` to False
+        super().__init__(False, *args, **kwargs)
+
+    @property
+    def sizes(self) -> list:
+        """"""
+        return {e: len(x) for e, x in zip(self.entities, self._datasets)}
+
+    def __len__(self) -> int:
+        """Total count of elements in the datset"""
+        return sum(self.sizes.values())
+
+    def __getitem__(
+        self, index_entity: Tuple[Union[str, int], int]
+    ) -> torch.Tensor:
```

Comment:
I'm not sure the torch data loader understands this notation? We may have to implement a custom loader for that case?
It's a bit unintuitive that you always have to pass a tuple here like `dataset['monomer',0]`. 
Instead, the dataset could be accessible normally (with 0 to N) and the first k samples point to monomers, the next l to catalyst etc and you could always return  an ID/text tensor telling you the entity. If you selectively want to fetch one type of entities you could have a `set_entity` function that adapts the `getitem` to only return samples of the desired entity.

If you want to keep the current architecture,  think at least a warning/error should be raised if it is indexed with a single int instead of a tuple.

Reply:
Okay, I'll check that. Just for the context, I needed a helper class to load data for the synthetic polymers, so I decided to reuse the `PolymerDatase`  but adapt it to my particular use case. I can get rid of the getitem method and use a custom method to retrieve data. I did not really have the `DataLoader` logic in mind. I just wanted my code to be consistent with the loading pipeline we have.

Reply:
I'm fine with it either way! But yea, in case the current syntax of getitem is incompatible with DataLoader, it may be better to go with a custom method for data retrieval and adjust the getitem to sth along the lines of what I wrote above. In that way, we could use the DataLoader out of the box whenever it becomes necessary.

Reply:
Sounds good to me

Reply:
I switched to the approach you mentioned in the first comment.